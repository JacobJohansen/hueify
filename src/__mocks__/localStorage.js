module.exports = global.localStorage = {
  store: {},
  setItem (key, value) {
    return (this.store[key] = value)
  },
  getItem (key) {
    return this.store[key] || null
  }
}
