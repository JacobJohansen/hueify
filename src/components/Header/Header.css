header {
  --position: 0%;
  --width: 33.33333%;
  position: absolute;
  width: 100%;
  height: var(--header-height);
  background: var(--header-bg);
  color: black;
  line-height: var(--header-height);
  border-bottom: var(--border);
  z-index: 2;
}

header ul {
  display: flex;
  flex-wrap: nowrap;
  padding: 0;
  margin: 0;
}

header::after {
  content: '';
  display: block;
  width: var(--width);
  margin-top: -2px;
  height: 2px;
  background: var(--blue);
  transition: transform .2s;
  transform: translateX(var(--position));
}

header.groups { --position: 100%; }
header.settings { --position: 200%; }

header ul li {
  width: var(--width);
  list-style: none;
  text-align: center;
  cursor: pointer;
}

header ul li.active { color: var(--blue); }

header {
  color: var(--color);
  transition: transform .2s;
}
