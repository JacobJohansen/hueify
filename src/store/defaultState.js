export default {
  view: 'Lights',
  bridge: {},
  lights: [],
  groups: [],
  settings: {
    host: null
  }
}
