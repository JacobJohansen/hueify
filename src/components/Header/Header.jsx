import { Component } from 'preact'
import store, { subscribe } from 'store'
import setView from 'store/actions/setView'
import './header.css'

class Header extends Component {
  /**
   * Switch to a new view
   * @param {String} view
   */
  setView (view) {
    this.context.route(view)

    setTimeout(() => (
      store.dispatch(setView(view))
    ), 10)
  }

  /**
   * Check if current view matches tab
   * @param {String} view
   * @return {String} className
   */
  isView (view) {
    return this.props.view === view ? 'active' : ''
  }

  /**
   * Render header component
   * @return {JSX}
   */
  render () {
    return (
      <header className={this.props.view}>
        <ul>
          <li
            onClick={this.setView.bind(this, 'lights')}
            className={this.isView('lights')}
          >
            Lights
          </li>
          <li
            onClick={this.setView.bind(this, 'groups')}
            className={this.isView('groups')}
          >
            Groups
          </li>
          <li
            onClick={this.setView.bind(this, 'settings')}
            className={this.isView('settings')}
          >
            Settings
          </li>
        </ul>
      </header>
    )
  }
}

export default subscribe(Header)
