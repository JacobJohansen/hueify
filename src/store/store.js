/* global JSON, localStorage */
import { createStore } from 'redux'
import defaultState from './defaultState'

/**
 * Create a new redux store
 * @type {Redux}
 */
const store = createStore((
  state = (JSON.parse(localStorage.getItem('store')) || defaultState),
  action
) => {
  switch (action.type) {
    case 'SET_VIEW':
      return { ...state, view: action.view }
    case 'SET_HOST':
      return { ...state, settings: { host: action.host } }
    case 'SET_BRIDGE':
      return { ...state, bridge: action.bridge }
    case 'SET_LIGHTS':
      return { ...state, lights: action.lights }
    case 'SET_GROUPS':
      return { ...state, groups: action.groups }
    default:
      return { ...state }
  }
})

export default store

/**
 * Save state to localStorage on changes
 */
store.subscribe(() => {
  localStorage.setItem('store', JSON.stringify(store.getState()))
})
