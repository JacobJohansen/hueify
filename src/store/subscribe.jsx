import { Component } from 'preact'
import store from 'store'

export default (WrappedComponent, getState) => {
  if (getState === undefined) {
    getState = store => store.getState()
  }

  return class extends Component {
    subscription = null
    state = getState(store, this.props)

    componentDidMount () {
      this.subscription = store.subscribe(this.handleUpdate.bind(this))
    }

    componentWillUnmount () {
      this.subscription()
    }

    handleUpdate () {
      this.setState(getState(store, this.props))
    }

    render () {
      return <WrappedComponent {...this.state} {...this.props} />
    }
  }
}
