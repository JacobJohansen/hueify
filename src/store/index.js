import store from './store'
import subscribe from './subscribe'
import defaultState from './defaultState'

/**
 * Export default state and the subscribe HOC
 */
export {
  defaultState,
  subscribe
}

/**
 * Export the redux store
 */
export default store
