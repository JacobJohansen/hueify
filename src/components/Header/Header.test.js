/* eslint-env jest */
import '../../__mocks__/localStorage'
import { h, render } from 'preact'
import { Views } from 'preact-views'
import { default as renderToString } from 'preact-render-to-string'
import store from 'store'
import Header from './Header'

global.h = h

test('Should render the header component', () => {
  const vdom = renderToString(<Header />)
  expect(vdom).toMatchSnapshot()
})

test('Header should dispatch view change if a tab is clicked', done => {
  render(<Views view='header'><Header name='header' /></Views>, document.body)
  store.subscribe(done)
  document.querySelectorAll('li')[1].click()
})
