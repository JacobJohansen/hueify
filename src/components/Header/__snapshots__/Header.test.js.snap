// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the header component 1`] = `"<header class=\\"lights\\"><ul><li class=\\"active\\">Lights</li><li class>Groups</li><li class>Settings</li></ul></header>"`;
